### PR TITLE
Improve the compile and flash subcommands

### PR DESCRIPTION
**Pull request by skullydazed**
## Description

This adds two new features to `qmk compile` and `qmk flash`, as well as enhancing JSON keymaps so they don't pollute the tree:

* `--clean` flag: remove the build directory for a keyboard before compiling
* `--env` flag: pass environment variables down to make
* JSON keymaps no longer write to the keymaps directory, but instead write into the `$(KEYMAP_OUTPUT)/src` directory.
* Make flake8 happy with unrelated files

## Types of Changes

- [x] Core
- [x] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: lib/python/qmk/commands.py [hunk 1]
```diff
@@ -52,13 +63,52 @@ def compile_configurator_json(user_keymap, bootloader=None):
...
+    keymap_c.write_text(c_text)
 
     # Return a command that can be run to make the keymap and flash if given
-    if bootloader is None:
-        return create_make_command(user_keymap['keyboard'], user_keymap['keymap'])
-    return create_make_command(user_keymap['keyboard'], user_keymap['keymap'], bootloader)
+    verbose = 'true' if cli.config.general.verbose else 'false'
+    color = 'true' if cli.config.general.color else 'false'
+    make_command = [
+        'make',
```
Comment (tzarc) [inline lib/python/qmk/commands.py:82]: can we get some sort of parallel build support here?

Comment (tzarc) [inline lib/python/qmk/commands.py:82]: should probably allow for overriding the executable -- `gmake`, `remake` etc.

Comment (skullydazed) [reply lib/python/qmk/commands.py]: a12168f89 adds this.

Comment (skullydazed) [reply lib/python/qmk/commands.py]: added in bae2c9270

**Review by tzarc — APPROVED** [review]
Weak approval.

**Comment by ReVanTis** [comment]
If I may comment here, I think I uploaded a pull request several weeks ago regarding passing "jobs" parameters here:
https://github.com/qmk/qmk_firmware/pull/10919

The conclusion from @skullydazed and @zvecr was not to implement the mechanism to pass MAKEFLAGS.
Just out of curiosity, has the conclusion changed since I saw the qmk command behavior has changed to respect "-j" parameters?

**Comment by skullydazed** [comment]
Since we had that discussion some of the direction has changed. At the time `qmk compile` operated strictly by interfacing with `Makefile`. We have since determined that doing so is too restrictive for some current and future use cases (compiling JSON files, supporting external keymaps/keyboards) and so we are moving towards bypassing `Makefile` instead. This necessitates supporting more options in `qmk compile`.

**Pull request merged**